### PR TITLE
Ryan 1744 taskcandidaterestrictions

### PR DESCRIPTION
**Pull request by ryandawsonuk**
for Activiti/Activiti#1744

### CODE REVIEW

File: activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/TaskCandidateGroupAddedEventHandler.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        TaskCandidateGroupAddedEvent taskCandidateGroupAddedEvent = (TaskCandidateGroupAddedEvent) event;
+        TaskCandidateGroup taskCandidateGroup = taskCandidateGroupAddedEvent.getTaskCandidateGroup();
+        
+        // not going to look up task as candidate can be created before task
+
+
+        // Persist into database
+        try {
+            taskCandidateGroupRepository.save(taskCandidateGroup);
+        } catch(Exception cause) {
```
Comment (erdemedeiros) [inline activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/TaskCandidateGroupAddedEventHandler.java]: Maybe better to catch a `Throwable` here.

File: activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/TaskCandidateGroupRemovedEventHandler.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+    @Override
+    public void handle(ProcessEngineEvent event) {
+        TaskCandidateGroupRemovedEvent taskCandidateGroupRemovedEvent = (TaskCandidateGroupRemovedEvent) event;
+        TaskCandidateGroup taskCandidateGroup = taskCandidateGroupRemovedEvent.getTaskCandidateGroup();
+
+        // remove from database
+        try {
+            taskCandidateGroupRepository.delete(taskCandidateGroup);
+        } catch(Exception cause) {
```
Comment (erdemedeiros) [inline activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/TaskCandidateGroupRemovedEventHandler.java]: Catch `Throwable`.

File: activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/TaskCandidateUserAddedEventHandler.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        TaskCandidateUserAddedEvent taskCandidateUserAddedEvent = (TaskCandidateUserAddedEvent) event;
+        TaskCandidateUser taskCandidateUser = taskCandidateUserAddedEvent.getTaskCandidateUser();
+        
+        // not going to look up task as candidate can be created before task
+
+
+        // Persist into database
+        try {
+            taskCandidateUserRepository.save(taskCandidateUser);
+        } catch(Exception cause) {
```
Comment (erdemedeiros) [inline activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/TaskCandidateUserAddedEventHandler.java]: Throwable

File: activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/TaskCandidateUserRemovedEventHandler.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+    @Override
+    public void handle(ProcessEngineEvent event) {
+        TaskCandidateUserRemovedEvent taskCandidateUserRemovedEvent = (TaskCandidateUserRemovedEvent) event;
+        TaskCandidateUser taskCandidateUser = taskCandidateUserRemovedEvent.getTaskCandidateUser();
+
+        // remove from database
+        try {
+            taskCandidateUserRepository.delete(taskCandidateUser);
+        } catch(Exception cause) {
```
Comment (erdemedeiros) [inline activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/TaskCandidateUserRemovedEventHandler.java]: Throwable

File: activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/AdminController.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+;
+
+@RestController
+@RequestMapping(value = "/admin/v1", produces = MediaTypes.HAL_JSON_VALUE)
```
Comment (erdemedeiros) [inline activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/AdminController.java]: To be updated after merging Media types PoC

File: activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/AdminController.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+public class AdminController {
+
+    private final TaskRepository taskRepository;
+
+    private TaskResourceAssembler taskResourceAssembler;
+
+    private PagedResourcesAssembler<Task> pagedResourcesAssembler;
```
Comment (erdemedeiros) [inline activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/AdminController.java]: To be updated after merging MediaTypes PoC

File: activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/AdminController.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+;
+
+@RestController
+@RequestMapping(value = "/admin/v1", produces = MediaTypes.HAL_JSON_VALUE)
+public class AdminController {
```
Comment (erdemedeiros) [inline activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/AdminController.java]: I'm wondering if it wouldn't be better to split this class into `AdminTaskController` and `AdminVariableController`. The class content is very small, I know, but it has 6 constructor's parameters and the first 3 are used only in the first method and the last 3 are used only in the second method. What do you think?

File: activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/security/TaskLookupRestrictionService.java [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+
+@Component
+public class TaskLookupRestrictionService {
+
+    @Autowired(required = false)
```
Comment (erdemedeiros) [inline activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/security/TaskLookupRestrictionService.java]: Prefer constructor injection.

File: activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/security/TaskLookupRestrictionService.java [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+import org.activiti.cloud.services.query.model.QVariable;
+import org.activiti.engine.UserGroupLookupProxy;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+
+@Component
+public class TaskLookupRestrictionService {
```
Comment (erdemedeiros) [inline activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/security/TaskLookupRestrictionService.java]: Missing unit test.

File: activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/security/VariableLookupRestrictionService.java [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+import org.springframework.stereotype.Component;
+
+/**
+ * This is present in case of a future scenario where we need to filter task or process instance variables more generally rather than per task or per proc.
+ */
+@Component
+public class VariableLookupRestrictionService {
+
+    @Autowired
+    private TaskLookupRestrictionService taskLookupRestrictionService;
```
Comment (erdemedeiros) [inline activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/security/VariableLookupRestrictionService.java]: Prefer constructor injection.

File: activiti-cloud-services-query/activiti-cloud-services-query-rest/src/test/java/org/activiti/cloud/services/query/events/handlers/VariableCreatedEventHandlerIT.java [hunk 11]
```diff
@@ -56,6 +54,12 @@
     static class Configuation {
     }
 
+
+    @After
```
Comment (erdemedeiros) [inline activiti-cloud-services-query/activiti-cloud-services-query-rest/src/test/java/org/activiti/cloud/services/query/events/handlers/VariableCreatedEventHandlerIT.java]: Well done! This save us a lot of time!

File: activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/ProcessInstanceVariableController.java [hunk 12]
```diff
@@ -49,9 +55,19 @@ public ProcessInstanceVariableController(VariableResourceAssembler variableResou
...
     public PagedResources<VariableResource> getVariables(@PathVariable String processInstanceId,
+                                                         @QuerydslPredicate(root = Variable.class) Predicate predicate,
                                                          Pageable pageable,
                                                          PagedResourcesAssembler<Variable> pagedResourcesAssembler) {
-        Page<Variable> variables = variableRepository.findAll(QVariable.variable.processInstanceId.eq(processInstanceId), pageable);
+
+        QVariable variable = QVariable.variable;
+        BooleanExpression expression = variable.processInstanceId.eq(processInstanceId);
+
+        if(predicate != null){
```
Comment (erdemedeiros) [inline activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/ProcessInstanceVariableController.java]: Missing else condition. If the predicate is null we still want to apply the `expression` condition.

Comment (erdemedeiros) [reply activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/ProcessInstanceVariableController.java]: Fixed with the last commit.

File: activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/security/TaskLookupRestrictionService.java (hunk 8 above)
Comment (ryandawsonuk) [reply activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/security/TaskLookupRestrictionService.java]: Generally I do too with optional attributes it's not so great. These can be handled with the Optional class as per https://stackoverflow.com/questions/42135102/how-to-set-autowired-constructor-params-as-required-false-individually but it is not so readable and doesn't work with @InjectMocks like in https://github.com/Activiti/activiti-cloud-query-service/blob/69c537861ad48753d66712130f51e110fd8990c8/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/test/java/org/activit/cloud/services/security/SecurityPoliciesApplicationServiceTest.java#L37 so I'd then have to change that test class to use setters to inject the mocks.

File: activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/security/VariableLookupRestrictionService.java (hunk 10 above)
Comment (ryandawsonuk) [reply activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/security/VariableLookupRestrictionService.java]: As previous

File: activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/TaskCandidateGroupRemovedEventHandler.java (hunk 2 above)
Comment (ryandawsonuk) [reply activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/TaskCandidateGroupRemovedEventHandler.java]: I think Exception is more appropriate in this case

File: activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/security/TaskLookupRestrictionService.java (hunk 8 above)
Comment (erdemedeiros) [reply activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/security/TaskLookupRestrictionService.java]: 👍

**Pull request merged**